e2e storage: embedded csi mock driver

pohly opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This replaces embedding of JavaScript code into the mock driver that
runs inside the cluster with Go callbacks which run inside the
e2e.test suite itself. In contrast to the JavaScript hooks, they have
direct access to all parameters and can fabricate arbitrary responses,
not just error codes.
    
Because the callbacks run in the same process as the test itself, it
is possible to set up two-way communication via shared variables or
channels. This opens the door for writing better tests. Some of the
existing tests that poll mock driver output could be simplified, but
that can be addressed later.
    
For now, the embedded mock driver emulates the behavior of the
original one as much as possible, including recording calls via JSON,
to minimize the amount of changes.

**Special notes for your reviewer**:

This approach depends on:
- updated csi-test: https://github.com/kubernetes-csi/csi-test/pull/314
- updated node-driver-registrar: https://github.com/kubernetes-csi/node-driver-registrar/pull/124
- port forwarding, which is currently not required for conformance, mostly (only?) because the corresponding test haven't been marked as `[Conformance]` yet

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

pohly committed:
e2e: import csi-test mock driver

This is a verbatim copy of the corresponding files in csi-test v4.0.2.

They'll be modified in future commits to make the code usable when
embedded in e2e.test. Some of those changes may be worthwhile
backporting to csi-test, but this is uncertain at this time.

pohly committed:
e2e storage: use embedded mock CSI driver

This replaces embedding of JavaScript code into the mock driver that
runs inside the cluster with Go callbacks which run inside the
e2e.test suite itself. In contrast to the JavaScript hooks, they have
direct access to all parameters and can fabricate arbitrary responses,
not just error codes.

Because the callbacks run in the same process as the test itself, it
is possible to set up two-way communication via shared variables or
channels. This opens the door for writing better tests. Some of the
existing tests that poll mock driver output could be simplified, but
that can be addressed later.

For now, only tests using hooks use embedding. How gRPC calls are
retrieved is abstracted behind the CSIMockTestDriver interface, so
tests don't need to be modified when switching between embedding
and remote mock driver.

pohly commented:
/sig storage
/hold

pohly commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -761,7 +754,16 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
+						}
+						return nil, nil
+					}
+				}(),
```
The Go callbacks are obviously longer than the one-line JS code. That's partly because Go is more verbose (strict typing), partly because the callbacks are more flexible (more parameters).

Personally I prefer writing everything in one language with compile-time checks, but perhaps it would be worthwhile to support also simpler callbacks?

One idea that I had was to turn each hook into a struct that accepts a full-blown callback with hook-specific parameters and a generic one that just takes `ctx context.Context, counter int`, with the counter maintained by code in the mock driver. That code then can also deal with concurrency better - in this code, access to `counter` is not protected against concurrent calls (a theoretic issue, but still not quite correct).

Would that be worthwhile?

jsafrane replied:
IMO, more verbose hooks are expected in golang and OK, the example above looks quite readable. And every hook can put a lock around their counter (or use atomic int) if they want. I don't see a big issue here.

pohly replied:
I've pushed a revised implementation where the hooks are implemented entirely inside the e2e.test binary in a generic way via a gRPC interceptor. The API towards the mock driver then becomes "replace normal interceptor with custom one", which is something that we can add to csi-test and then vendor into k/k. For now we still need a fork until we have a csi-test release without JavaScript dependency.

Inside the mock tests, hooks also became simpler thanks to a small helper function which provides the common code for counting method calls.

If we need more complex functionality, then more complex hooks can inspect the request and produce a response in addition to errors.

pohly commented on test/e2e/storage/drivers/csi-test/mock/service/service.go:
```diff
@@ -0,0 +1,326 @@
+type CreateSnapshotHook func(context.Context, *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error)
+type NodeStageHook func(context.Context, *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error)
+type NodeUnstageHook func(context.Context, *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error)
+type NodeGetVolumeStatsHook func(context.Context, *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error)
```
So far I have only converted those hooks that are actually needed. Without generics, adding hooks to a function is a bit of (tedious) cut-and-paste.

Should we aim for instrumenting all functions from the beginning or can that be added as needed? That would work because the tests and the driver are now in the same repo.

pohly replied:
As mentioned in https://github.com/kubernetes/kubernetes/pull/97069#discussion_r572213634, I've now pushed a revision that doesn't add any of these manually defined hooks.

pohly committed:
CSI mock driver: fix faulty error message

Caught by verify-typecheck.sh after importing the code into
Kubernetes:

ERROR(linux/arm): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:404:20: math.MaxUint32 (untyped int constant 4294967295) overflows int
ERROR(linux/arm): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:795:20: math.MaxUint32 (untyped int constant 4294967295) overflows int
ERROR(linux/386): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:404:20: math.MaxUint32 (untyped int constant 4294967295) overflows int
ERROR(linux/386): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:795:20: math.MaxUint32 (untyped int constant 4294967295) overflows int
ERROR(windows/386): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:404:20: math.MaxUint32 (untyped int constant 4294967295) overflows int
ERROR(windows/386):
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/service/controller.go:795:20:
math.MaxUint32 (untyped int constant 4294967295) overflows int

Instead of producing our own error message, we can show the original
value and the error from strconv.

fedebongio commented:
/remove-sig api-machinery

jsafrane commented on test/e2e/testing-manifests/storage-csi/mock/csi-mock-proxy.yaml:
```diff
@@ -0,0 +1,90 @@
+            name: registration-dir
+        - name: mock
+          # TODO: replace with official image which includes https://github.com/kubernetes-csi/csi-test/pull/314
+          image: pohly/mock-driver:2020-12-04-1
```
Does it make sense to build a separate csi-mock-proxy container (+github repo)? Because it does not share much with the mock driver.

pohly replied:
We could do that. But I'd prefer to avoid the overhead of managing a separate repo and image. As an intermediate solution would could keep the code in the csi-test repo and introduce a `csi-proxy` command and image, but even that IMHO doesn't offer much actual advantages besides "it's a bit cleaner".

jsafrane replied:
Maybe a crazy idea, can the node driver registrar register a `tcp://` socket of a CSI driver + can we deploy a generic TCP proxy (HA proxy?). We won't need any extra container on the nodes.

pohly replied:
It's not just kubelet, we also want to test with the actual sidecar containers, right? So we do need to run containers inside Kubernetes. The sidecars also could use TCP, but I doubt that this makes a difference: the proxy must establish a connection between *two* incoming TCP connection requests, one coming from kubelet or sidecar, the other coming from e2e.test outside of the cluster. I doubt that a generic proxy can do that.

It would be more natural to forward a listening port opened by the e2e.test binary into the cluster, but that is not supported by the port-forwarding feature of the apiserver, at least as far as I know - it just supports incoming connections, not outgoing ones.

pohly replied:
Yet another possibility: create a `csi-proxy` command and publish it in the same `mock-driver` image. Hmm, this might be even more work than adding a second image. A second image should already be supported by the csi-release-tools (albeit not tested, I think) whereas a second command per image isn't.

jsafrane replied:
Ack, let's keep it in the mock driver for now. We can always split it later  if it becomes too complicated to maintain a proxy and a driver in the same binary.

pohly replied:
> Ack, let's keep it in the mock driver for now.

Can you then approve https://github.com/kubernetes-csi/csi-test/pull/314? That's the implementation of that approach.

jsafrane commented on test/e2e/testing-manifests/storage-csi/mock/csi-mock-proxy.yaml:
```diff
@@ -0,0 +1,90 @@
+              name: socket-dir
+        # The busybox container is needed for running shell commands which
+        # test for directories or create them.
+        - name: busybox
```
If I understood it correctly, this container needs to privileged, because it's going to create directories in `/var/lib/kubelet/pods/...`. On machines with SELinux, only privileged containers may modify it.

pohly replied:
Correct, will fix this.

pohly replied:
Done.

jsafrane replied:
I still see it as unprivileged

pohly replied:
Where the heck did my change go? Ah, into the wrong container (external-provisioner above). Fixed.

jsafrane reviewed (commented):
This is actual log of a test:
```
STEP: Waiting for expected CSI calls
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.433147       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.433204       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.433264       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.607681       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 243 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.607974       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 146 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41274: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.608119       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.608145       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.608191       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.782940       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 480 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:52.783039       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 109 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41276: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.338077       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.338133       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.338209       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.629286       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.629331       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.629383       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.629486       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 243 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.629530       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 146 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41278: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.820147       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 480 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:53.820241       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 109 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41280: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:54.873693       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:54.873746       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:54.873807       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.177067       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 243 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.177142       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 146 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41282: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.177430       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.177472       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.177525       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.371008       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 480 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:55.371113       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 109 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41284: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.404442       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.404497       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.404560       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.928928       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.928976       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.929026       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.965125       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 243 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:57.965203       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 146 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41288: use of closed network connection
STEP: Waiting for pod to be running
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.465062       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 510 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.465139       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 124 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41300: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.476859       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.476902       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.476957       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.659988       1 proxy.go:89] proxy established a new connection between /csi/csi.sock and tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.660033       1 proxy.go:127] starting to copy tcp://:9000 -> /csi/csi.sock
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.660149       1 proxy.go:127] starting to copy /csi/csi.sock -> tcp://:9000
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.660247       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 243 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:36:59.660288       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 146 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41342: use of closed network connection
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:37:03.694937       1 proxy.go:133] done copying /csi/csi.sock -> tcp://:9000: 646 bytes, <nil>
csi-mockplugin-0/mock@ip-1..rnal: I0128 14:37:03.695017       1 proxy.go:133] done copying tcp://:9000 -> /csi/csi.sock: 124 bytes, read tcp 127.0.0.1:9000->127.0.0.1:41344: use of closed network connection
STEP: Deleting the previously created pod
```

As you can see, it's not very informative - lot of copying of messages between sockets, but there is no content of the messages anywhere. It's probably impossible to parse the TCP stream in the proxy and print the requests/responses there. Would it be possible to log them somewhere in the mock driver in k/k?

pohly commented:
> As you can see, it's not very informative - lot of copying of messages between sockets, but there is no content of the messages anywhere. It's probably impossible to parse the TCP stream in the proxy and print the requests/responses there.

The proxy.go logging is indeed a bit verbose. I'll check whether I can make that less verbose.

> Would it be possible to log them somewhere in the mock driver in k/k?

That's definitely doable. We can install the usual gRPC interceptors.

pohly committed:
generated files due to modified dependencies

pohly committed:
mock driver: replace logrus with klog

klog now has structured logging.

pohly committed:
e2e: reduce worker threads during mock testing

We don't need much concurrency and having too many worker threads has
one disadvantage (besides resource usage): when the sidecar looses the
connection to the CSI driver, it calls klog.Fatal, which prints all
gouroutines. This can lead to much output.

pohly committed:
mock driver fixes

pohly committed:
mock driver: fix no-op setDefaultCreds

The function must modify the content of the "creds" pointer, not the
pointer.

Found via hack/verify-staticcheck.sh after importing the code into
Kubernetes. It is uncertain whether this bug had any consequences.

pohly committed:
csi-test: import copyright clarification

The code originates in csi-test and this copyright change was made by
the original author in https://github.com/kubernetes-csi/csi-test/pull/324

pohly committed:
test: update CSI mock driver

The new release properly checks for existence of the staging
directory. The names of the volumes are now the same as in
https://github.com/kubernetes-csi/csi-test/blob/master/mock/example/deploy/csi-mock-driver-deployment.yaml

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

pohly committed:
e2e storage: simpler port forwarding

Instead of trying to use the client-go portforward package as-is it is
simpler to copy some code from it and then use the http stream
directly. That way we don't need to go through a local listening
socket and error handling and logging becomes simpler.

pohly committed:
e2e test: log gRPC calls in embedded CSI driver

It is useful to see all calls as they occur. The output format is
the more readable JSON representation.

pohly committed:
e2e test: relax CSI call expectations

NodeUnstageVolume and DeleteVolume are not necessarily ordered and in
practice, DeleteVolume was indeed encountered first after changing the
timing by embedding the CSI driver:

Kubernetes e2e suite: [sig-storage] CSI mock volume storage capacity exhausted, immediate binding expand_less 	7m34s
test/e2e/storage/csi_mock_volume.go:953
Feb 26 18:34:04.037: while waiting for all CSI calls
Unexpected error:
    <*errors.errorString | 0xc00322b9c0>: {
        s: "error waiting for expected CSI calls: Unexpected CSI call 3: expected NodeUnstageVolume (0), got DeleteVolume (0)",
    }
    error waiting for expected CSI calls: Unexpected CSI call 3: expected NodeUnstageVolume (0), got DeleteVolume (0)
occurred
test/e2e/storage/csi_mock_volume.go:1045

pohly committed:
e2e test: use one connection per stream

Sharing the same connection for multiple streams should have worked,
but ran into unexpected timeouts:

I0227 08:07:49.754263   80029 portproxy.go:109] container "mock" in pod csi-mock-volumes-4037-2061/csi-mockplugin-0 is running
E0227 08:07:49.779359   80029 portproxy.go:178] prepare forwarding csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: dialer failed: unable to upgrade connection: pod not found ("csi-mockplugin-0_csi-mock-volumes-4037-2061")
I0227 08:07:50.782705   80029 portproxy.go:109] container "mock" in pod csi-mock-volumes-4037-2061/csi-mockplugin-0 is running
I0227 08:07:50.809326   80029 portproxy.go:125] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: starting connection polling
I0227 08:07:50.909544   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #0, 0 open
I0227 08:07:50.912436   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #0
I0227 08:07:50.912503   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #0
I0227 08:07:50.913161   80029 portproxy.go:322] forward connection #0 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
E0227 08:07:50.913324   80029 portproxy.go:242] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: an error occurred connecting to the remote port: error forwarding port 9000 to pod 66662ea1ab30b4193dac0102c49be840971d337c802cc0c8bbc074214522bd13, uid : failed to execute portforward in network namespace "/var/run/netns/cni-c15e4e36-dad9-8316-c301-33af9dad5717": failed to dial 9000: dial tcp4 127.0.0.1:9000: connect: connection refused
I0227 08:07:50.913371   80029 portproxy.go:340] forward connection #0 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
W0227 08:07:50.913487   80029 server.go:669] grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams failed to receive the preface from client: EOF"
I0227 08:07:51.009519   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #1, 0 open
I0227 08:07:51.011912   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #1
I0227 08:07:51.011973   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #1
I0227 08:07:51.013677   80029 portproxy.go:322] forward connection #1 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:07:51.013720   80029 portproxy.go:340] forward connection #1 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
W0227 08:07:51.013794   80029 server.go:669] grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams failed to receive the preface from client: EOF"
E0227 08:07:51.017026   80029 portproxy.go:242] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: an error occurred connecting to the remote port: error forwarding port 9000 to pod 66662ea1ab30b4193dac0102c49be840971d337c802cc0c8bbc074214522bd13, uid : failed to execute portforward in network namespace "/var/run/netns/cni-c15e4e36-dad9-8316-c301-33af9dad5717": failed to dial 9000: dial tcp4 127.0.0.1:9000: connect: connection refused
I0227 08:07:51.109515   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #2, 0 open
I0227 08:07:51.111479   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #2
I0227 08:07:51.111519   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #2
I0227 08:07:51.209519   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #3, 1 open
I0227 08:07:51.766305   80029 csi.go:377] gRPC call: {"Method":"/csi.v1.Identity/Probe","Request":{},"Response":{"ready":{"value":true}},"Error":"","FullError":null}
I0227 08:07:51.768304   80029 csi.go:377] gRPC call: {"Method":"/csi.v1.Identity/GetPluginInfo","Request":{},"Response":{"name":"csi-mock-csi-mock-volumes-4037","vendor_version":"0.3.0","manifest":{"url":"https://k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock"}},"Error":"","FullError":null}
I0227 08:07:51.770494   80029 csi.go:377] gRPC call: {"Method":"/csi.v1.Identity/GetPluginCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Service":{"type":1}}},{"Type":{"VolumeExpansion":{"type":1}}},{"Type":{"Service":{"type":2}}}]},"Error":"","FullError":null}
I0227 08:07:51.772899   80029 csi.go:377] gRPC call: {"Method":"/csi.v1.Controller/ControllerGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":3}}},{"Type":{"Rpc":{"type":10}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":6}}},{"Type":{"Rpc":{"type":5}}},{"Type":{"Rpc":{"type":8}}},{"Type":{"Rpc":{"type":7}}},{"Type":{"Rpc":{"type":12}}},{"Type":{"Rpc":{"type":11}}},{"Type":{"Rpc":{"type":9}}}]},"Error":"","FullError":null}
I0227 08:08:21.209901   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:08:21.209980   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #3, 1 open
I0227 08:08:51.211522   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating data stream: Timeout occurred
I0227 08:08:51.211566   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #3, 1 open
I0227 08:08:51.213451   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #3
I0227 08:08:51.213498   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #3
I0227 08:08:51.309540   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #4, 2 open
I0227 08:08:52.215358   80029 portproxy.go:322] forward connection #3 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:08:52.215475   80029 portproxy.go:340] forward connection #3 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
I0227 08:09:21.310003   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:09:21.310086   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #4, 1 open
I0227 08:09:51.311854   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating data stream: Timeout occurred
I0227 08:09:51.311908   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #4, 1 open
I0227 08:09:51.314415   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #4
I0227 08:09:51.314497   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #4
I0227 08:09:51.409527   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #5, 2 open
I0227 08:09:52.326203   80029 portproxy.go:322] forward connection #4 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:09:52.326277   80029 portproxy.go:340] forward connection #4 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
I0227 08:10:21.409892   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:10:21.409954   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #5, 1 open
I0227 08:10:51.411455   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating data stream: Timeout occurred
I0227 08:10:51.411557   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #5, 1 open
I0227 08:10:51.413229   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #5
I0227 08:10:51.413274   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #5
I0227 08:10:51.509508   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #6, 2 open
I0227 08:10:52.414862   80029 portproxy.go:322] forward connection #5 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:10:52.414931   80029 portproxy.go:340] forward connection #5 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
I0227 08:11:21.509879   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:11:21.509934   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #6, 1 open
I0227 08:11:51.511519   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating data stream: Timeout occurred
I0227 08:11:51.511568   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #6, 1 open
I0227 08:11:51.513519   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #6
I0227 08:11:51.513571   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #6
I0227 08:11:51.609504   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #7, 2 open
I0227 08:11:52.517799   80029 portproxy.go:322] forward connection #6 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:11:52.517918   80029 portproxy.go:340] forward connection #6 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
I0227 08:12:21.609856   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:12:21.609909   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #7, 1 open
I0227 08:12:51.611494   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating data stream: Timeout occurred
I0227 08:12:51.611555   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #7, 1 open
I0227 08:12:51.613289   80029 portproxy.go:155] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: created a new connection #7
I0227 08:12:51.613343   80029 portproxy.go:286] forward listener for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: got a new connection #7
I0227 08:12:51.709535   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #8, 2 open
I0227 08:12:52.615858   80029 portproxy.go:322] forward connection #7 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: remote side closed the stream
I0227 08:12:52.615989   80029 portproxy.go:340] forward connection #7 for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: closing our side
W0227 08:12:52.616116   80029 server.go:669] grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams failed to receive the preface from client: EOF"
I0227 08:13:21.709934   80029 portproxy.go:151] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: no connection: error creating error stream: Timeout occurred
I0227 08:13:21.709997   80029 portproxy.go:148] port forwarding for csi-mock-volumes-4037-2061/csi-mockplugin-0:9000: trying to create a new connection #8, 1 open
Feb 27 08:13:30.916: FAIL: Failed to register CSIDriver csi-mock-csi-mock-volumes-4037
Unexpected error:
    <*errors.errorString | 0xc002666220>: {
        s: "error waiting for CSI driver csi-mock-csi-mock-volumes-4037 registration on node kind-worker2: timed out waiting for the condition",
    }
    error waiting for CSI driver csi-mock-csi-mock-volumes-4037 registration on node kind-worker2: timed out waiting for the condition
occurred

pohly commented:
This is now working as intended. We agreed to transition towards error injection via the embedded CSI mock driver, with further refactoring of the mock driver to follow later.

Can I get this reviewed and merged?

@kubernetes/dep-approvers 
@jsafrane

pohly commented:
/retest

pohly commented:
/retest

jsafrane commented on test/utils/image/manifest.go:
```diff
@@ -30,6 +30,7 @@ import (
 
 // RegistryList holds public and private image registries
 type RegistryList struct {
+	AlpineRegistry          string `yaml:"alpineRegistry"`
```
Where is this registry used?

pohly replied:
Ooops. This can be removed. It was used when doing the proxying via an alpine socat image, but then that got replaced by the proxy code in csi-test.

jsafrane commented on test/e2e/storage/drivers/proxy/portproxy.go:
```diff
@@ -0,0 +1,330 @@
+
+// Addr contains all relevant parameters for a certain port in a pod.
+// The container must be running before connections are attempted.
+type Addr struct {
```
It looks like it implements some interface, please add a check for it.

pohly replied:
net.Addr, added.

jsafrane commented on test/e2e/storage/drivers/proxy/portproxy.go:
```diff
@@ -0,0 +1,330 @@
+	}
+}
+
+type connection struct {
```
It looks like it implements some interface, please add a check for it.

pohly replied:
net.Connection, added.

jsafrane commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -416,25 +598,97 @@ func (m *mockCSIDriver) PrepareTest(f *framework.Framework) (*storageframework.P
+	return
+}
+
+func (m *mockCSIDriver) GetCalls() ([]MockCSICall, error) {
```
I still see some tests run with not embedded CSI drivers, is there plan to move them towards it? In such case a TODO comment to store intercepted calls above and returning them here would save us some ugly JSON parsing.

pohly replied:
Not for now. In some earlier revision of the PR I ran all tests with embedded mock driver and just captured calls directly in memory. But then @msau42 pointed out that tests that are meant to become part of conformance can't depend on the port forwarding API in Kubernetes because that itself currently isn't required for conformance.

My plan for now is to work a bit more with this new approach and get it exercised regularly as part of E2E testing to be confident with it.

Perhaps I should simply submit the existing port forwarding tests for inclusion in conformance testing.... I asked in https://kubernetes.slack.com/archives/C78F00H99/p1606322090151300 why it wasn't included but didn't get much feedback.

pohly commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -155,12 +149,29 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			FSGroupPolicy:       tp.fsGroupPolicy,
 		}
 
+		// At the moment, only tests which need hooks are
+		// using the embedded CSI mock driver. The rest run
+		// the driver inside the cluster although they could
+		// changed to use embedding merely by setting
+		// driverOpts.embedded to true.
```
Yes, and that's what I meant here: all usage of JavaScript gets replaced by hooks, and only those tests then run with the embedded CSI mock driver.

pohly commented on test/e2e/storage/drivers/proxy/io.go:
```diff
@@ -0,0 +1,79 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+type PodDirIO struct {
```
Added.

pohly commented:
@jsafrane: PR updated and rebased to remove the Bazel build files (yeah!).

pohly commented:
/retest

jsafrane commented:
/lgtm
/approve
(at least the storage parts)

dims commented:
we already vendor and use github.com/golang/protobuf 1.4.3, so the dependency changes LGTM

/approve
/lgtm

/assign @smarterclayton @liggitt 

Jordan, Clayton, one of you have to look at this lots of commits touching lots of branches.

pohly commented:
> Jordan, Clayton, one of you have to look at this lots of commits touching lots of branches.

All except for `go.mod` and `vendor/module.txt` are inside `test/e2e/storage` though. It looks like those two files still need approval.

liggitt commented:
/approve

for go.mod change

pohly commented:
/hold cancel

Ready for merging...

pohly commented:
/retest
